src/projections/s2.cpp

```diff
@@ -0,0 +1,386 @@
+    x = coslat * coslon;
+    y = coslat * sinlon;
+    z = sinlat;
+
+    PJ_XYZ spherePoint {x, y, z};
+    PJ_XY uvCoords;
+
+    ValidFaceXYZtoUV(Q->face, spherePoint, &uvCoords.x, &uvCoords.y);
+    double s = UVtoST(uvCoords.x, Q->UVtoST);
+    double t = UVtoST(uvCoords.y, Q->UVtoST);
+    return {s, t};
+}
+
+static PJ_LP s2_inverse (PJ_XY xy, PJ *P) {     
+    PJ_LP lp = {0.0,0.0};
+    struct pj_opaque *Q = static_cast<struct pj_opaque*>(P->opaque);
+
+    // Do the S2 projections to get from s,t to u,v to x,y,z
+    double u = STtoUV(xy.x, Q->UVtoST);
+    double v = STtoUV(xy.y, Q->UVtoST);
+
+    PJ_XYZ sphereCoords;
+    UVtoSphereXYZ(Q->face, u, v, &sphereCoords);
+    double q = sphereCoords.x;
+    double r = sphereCoords.y;
+    double s = sphereCoords.z;
+
+    // Get the spherical angles from the x y z
+    lp.phi = acos(-s) - M_HALFPI;
+    lp.lam = atan2(r, q);
+    
+    /* Apply the shift from the sphere to the ellipsoid as described
+     * in [LK12]. */
+    if (P->es != 0.0) {
+        int invert_sign;
+        double tanphi, xa;
+        invert_sign = (lp.phi < 0.0 ? 1 : 0);
+        tanphi = tan(lp.phi);
+        xa = P->b / sqrt(tanphi * tanphi + Q->one_minus_f_squared);
+        lp.phi = atan(sqrt(P->a * P->a - xa * xa) / (Q->one_minus_f * xa));
+        if (invert_sign) {
+            lp.phi = -lp.phi;
+        }
+    }
+
+    return lp;
+}
+
+PJ *PROJECTION(s2) {
+    struct pj_opaque *Q = static_cast<struct pj_opaque*>(calloc (1, sizeof (struct pj_opaque)));
+    if (nullptr==Q)
+        return pj_default_destructor (P, PROJ_ERR_OTHER /*ENOMEM*/);
+    P->opaque = Q;
+
+    /* Determine which UVtoST function is to be used */
+    PROJVALUE maybeUVtoST = pj_param(P->ctx, P->params, "sUVtoST");
+    if (nullptr != maybeUVtoST.s) {
+        try {
+            Q->UVtoST = stringToS2ProjectionType.at(maybeUVtoST.s);
+        } catch (std::out_of_range) {
+            proj_log_error(P, _("Invalid value for s2 parameter: should be linear, quadratic, tangent, or none."));
+            return pj_default_destructor (P, PROJ_ERR_INVALID_OP_ILLEGAL_ARG_VALUE);
+	}
+    } else {
+	    Q->UVtoST = Quadratic;
+    }
+
+    P->left = PJ_IO_UNITS_RADIANS;
+    P->right = PJ_IO_UNITS_PROJECTED;
+    P->from_greenwich = -P->lam0;
+
+    P->inv = s2_inverse;
+    P->fwd = s2_forward;
+
+    /* Determine the cube face from the center of projection. */
+    if (P->phi0 >= M_HALFPI - M_FORTPI / 2.0) {
+        Q->face = FACE_TOP;
+    } else if (P->phi0 <= -(M_HALFPI - M_FORTPI / 2.0)) {
+        Q->face = FACE_BOTTOM;
+    } else if (fabs(P->lam0) <= M_FORTPI) {
+        Q->face = FACE_FRONT;
+    } else if (fabs(P->lam0) <= M_HALFPI + M_FORTPI) {
+        Q->face = (P->lam0 > 0.0 ? FACE_RIGHT : FACE_LEFT);
+    } else {
+        Q->face = FACE_BACK;
+    }
+    /* Fill in useful values for the ellipsoid <-> sphere shift
+     * described in [LK12]. */
+    if (P->es != 0.0) {
+        Q->a_squared = P->a * P->a;
```

Comment:
I don't see a_squared to be used at all currently, but there's a P->a * P->a occurrence at line 330 that could use it